modules/lang-painless/src/main/java/org/elasticsearch/painless/node/AExpression.java

```diff
@@ -105,36 +49,51 @@
     /**
      * Standard constructor with location used for error tracking.
      */
-    AExpression(int indentifier, Location location) {
-        super(indentifier, location);
+    AExpression(int identifier, Location location) {
+        super(identifier, location);
     }
 
     /**
      * Checks for errors and collects data for the writing phase.
      */
-    Output analyze(ClassNode classNode, SemanticScope semanticScope, Input input) {
+    Output analyze(ClassNode classNode, SemanticScope semanticScope) {
         throw new UnsupportedOperationException();
     }
 
     /**
      * Checks for errors and collects data for the writing phase. Adds additional, common
      * error checking for conditions related to static types and partially constructed static types.
      */
-    static Output analyze(AExpression expression, ClassNode classNode, SemanticScope semanticScope, Input input) {
-        Output output = expression.analyze(classNode, semanticScope, input);
+    static Output analyze(AExpression expression, ClassNode classNode, SemanticScope semanticScope) {
+        Output output = expression.analyze(classNode, semanticScope);
+
+        if (semanticScope.hasDecoration(expression, PartialCanonicalTypeName.class)) {
+            throw expression.createError(new IllegalArgumentException("cannot resolve symbol " +
+                    "[" + semanticScope.getDecoration(expression, PartialCanonicalTypeName.class).getPartialCanonicalTypeName() + "]"));
+        }
 
-        if (output.partialCanonicalTypeName != null) {
-            throw expression.createError(new IllegalArgumentException("cannot resolve symbol [" + output.partialCanonicalTypeName + "]"));
+        if (semanticScope.hasDecoration(expression, StaticType.class)) {
+            throw expression.createError(new IllegalArgumentException("value required: instead found unexpected type " +
+                    "[" + semanticScope.getDecoration(expression, StaticType.class).getStaticCanonicalTypeName() + "]"));
         }
 
-        if (output.isStaticType) {
-            throw expression.createError(new IllegalArgumentException("value required: " +
-                    "instead found unexpected type [" + PainlessLookupUtility.typeToCanonicalTypeName(output.actual) + "]"));
+        if (semanticScope.hasDecoration(expression, ValueType.class) == false) {
+            throw expression.createError(new IllegalStateException("value required: instead found no value"));
         }
 
         return output;
     }
 
+    // TODO: move this somewhere more appropriate
```

Comment:
Why is this inappropriate?

Reply:
This will be moved in a future PR because the whole point of this is to decouple the decorations from anything in the user tree. This violates that principle for now.